lte/gateway/python/integ_tests/s1aptests/test_stateless_multi_ue_mixedstate_mme_restart.py

```diff
@@ -0,0 +1,292 @@
+                        ]
+        num_of_steps = len(attach_steps)
+
+        tot_num_ues = num_ues_idle + num_ues_active + num_ues_attaching
+        self._s1ap_wrapper.configUEDevice(tot_num_ues)
+
+        idle_session_ips = []
+        # Prep attached UEs
+        for i in range(num_ues_idle + num_ues_active):
+            req = self._s1ap_wrapper.ue_req
+            print("************************* sending Attach Request for "
+                  "UE id ", req.ue_id)
+            attach = self._s1ap_wrapper._s1_util.attach(
+                req.ue_id, s1ap_types.tfwCmd.UE_END_TO_END_ATTACH_REQUEST,
+                s1ap_types.tfwCmd.UE_ATTACH_ACCEPT_IND,
+                s1ap_types.ueAttachAccept_t
+            )
+
+            addr = attach.esmInfo.pAddr.addrInfo
+            default_ip = ipaddress.ip_address(bytes(addr[:4]))
+            if i < num_ues_idle:
+                idle_session_ips.append(default_ip)
+            else:
+                self.dl_flow_rules[default_ip] = []
+
+            # Wait on EMM Information from MME
+            self._s1ap_wrapper._s1_util.receive_emm_info()
+            ue_ids.append(req.ue_id)
+
+        # Move first num_ues_idle UEs to idle state
+        for i in range(num_ues_idle):
+            print(
+                "************************* Sending UE context release request ",
+                "for UE id ", ue_ids[i])
+            # Send UE context release request to move UE to idle mode
+            ue_cntxt_rel_req = s1ap_types.ueCntxtRelReq_t()
+            ue_cntxt_rel_req.ue_Id = ue_ids[i]
+            ue_cntxt_rel_req.cause.causeVal = (
+                gpp_types.CauseRadioNetwork.USER_INACTIVITY.value
+            )
+            self._s1ap_wrapper.s1_util.issue_cmd(
+                s1ap_types.tfwCmd.UE_CNTXT_REL_REQUEST, ue_cntxt_rel_req
+            )
+            response = self._s1ap_wrapper.s1_util.get_response()
+            self.assertEqual(
+                response.msg_type, s1ap_types.tfwCmd.UE_CTX_REL_IND.value
+            )
+
+        tId = {}
+        # start attach procedures for the remaining UEs
+        for i in range(num_ues_attaching):
+            req = self._s1ap_wrapper.ue_req
+            print("************************* Starting Attach procedure "
+                  "UE id ", req.ue_id)
+            # bring each newly attaching UE to the desired point during
+            # attach procedure before restarting mme service
+            ue_ids.append(req.ue_id)
+            for step in range(stateof_ues_in_attachproc_before_restart[i]):
+                if attach_steps[step] == self.exec_sec_mode_complete_step:
+                    tId[req.ue_id] = attach_steps[step](req.ue_id)
+                elif attach_steps[step] == self.exec_esm_inf_req_step:
+                    attach_steps[step](req.ue_id,tId[req.ue_id])
+                else:
+                    attach_steps[step](req.ue_id)
+
+
+        print("************************* Restarting MME service on",
```

Comment:
Can replace lines 209-215 with the newly added function `restart_mme_and_wait` in s1ap_utils.py